sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java

```diff
@@ -1089,6 +1098,8 @@ public void process(
                           initializedDatasetService,
                           initializedWriteStreamService,
                           pipelineOptions.as(BigQueryOptions.class)));
+      Lineage.getSinks()
```

Comment:
TableUrn has format "projects/%s/datasets/%s/tables/%s". We should be able to recover the table reference from urn so no need to change DestinationState? Otherwise could the change on DestinationState breaking upgrade compatibility?

Reply:
I'd prefer to keep the structured reference rather than re-parse it. Update should not be an issue as it's never serialized, it's just used locally.

Reply:
TableDestination can have null projectId that was why "tableDestination1.getTableUrn(bigQueryOptions)" was passed in here, which has a couple of fallback logics to resolve the full destination.

A potential issue for the current change if the sink is in a different project set by pipeline option, the metrics could report to a wrong destination

However I also notice that Lineage is per project basis. How to handle cross project observability may be an open question.

If you think this is fine, the change is good for now

Reply:
Ah, yes, this is a very good point. I'm taking BQ options as an option to this helper as well to ensure we have it everywhere. Thanks for catching this.